caller_test.go

```diff
@@ -0,0 +1,153 @@
+package logrus
+
+import (
+	"bufio"
+	"encoding/json"
+	"os"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type Output struct {
+	caller string `json:"caller"`
+	msg    string `json:"msg"`
+}
+
+func TestEntryTEXTFormatterCaller(t *testing.T) {
+	fileName := "caller.txt"
+	os.Remove(fileName)
+	f, err := os.OpenFile(fileName, os.O_CREATE|os.O_WRONLY, os.ModePerm)
+	assert.NoError(t, err)
+
+	logger := New()
+	logger.showCaller = true
+	logger.Level = DebugLevel
+	logger.Formatter = &TextFormatter{DisableColors: true}
+	logger.Out = f
+
+	check(logger)
+
+	f.Close()
+
+	f, err = os.Open(fileName)
+	defer f.Close()
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		ss := scanner.Text()
+		fields := strings.Fields(ss)
+		msg := ""
+		call := ""
+		for _, field := range fields {
+			if len(field) >= 7 && field[:7] == "caller=" {
+				msg = field[7:]
+			} else if len(field) >= 4 && field[:4] == "msg=" {
+				call = field[4:]
+			}
+		}
+		assert.Equal(t, msg, call)
+	}
+	err = os.Remove(fileName)
+	assert.NoError(t, err)
+}
+
+func TestEntryJSONFormatterCaller(t *testing.T) {
+	fileName := "caller.json"
+	os.Remove(fileName)
+	f, err := os.OpenFile(fileName, os.O_CREATE|os.O_WRONLY, os.ModePerm)
+	assert.NoError(t, err)
+
+	logger := New()
+	logger.showCaller = true
+	logger.Level = DebugLevel
+	logger.Formatter = &JSONFormatter{}
+	logger.Out = f
+
+	check(logger)
+
+	f.Close()
+
+	f, err = os.Open(fileName)
+	defer f.Close()
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		ss := scanner.Text()
+		var output Output
+		err = json.Unmarshal([]byte(ss), &output)
+		assert.NoError(t, err)
+		assert.Equal(t, output.caller, output.msg)
+	}
+	err = os.Remove(fileName)
+	assert.NoError(t, err)
+}
+
+func check(l *Logger) {
```

Comment:
this check should be rolled into a form of table testing, the intent is lost in the length of the function body, and it's hard to grasp what's being tested